python/ray/air/session.py

```diff
@@ -0,0 +1,167 @@
+import abc
+import logging
+from typing import Optional, Dict
+
+from ray.air.checkpoint import Checkpoint
+
+logger = logging.getLogger(__name__)
+
+
+class Session(abc.ABC):
+    """The canonical session interface that both Tune and Train session implements.
+
+    User can interact with this interface to get session information,
+    as well as reporting metrics and saving checkpoint.
+    """
+
+    @abc.abstractmethod
+    def report(self, metrics: Dict, checkpoint: Optional[Checkpoint] = None) -> None:
+        """Report metrics and optionally save checkpoint.
+
+        TODO(xwjiang): To be able to say the following without racing.
+        By the return of this method, the checkpoint is guaranteed to be saved onto
```

Comment:
@krfricke 
So I think there is an issue as upload to cloud is non-blocking. Consider if I do something like

mode.save("some_dir")
checkpoint = Checkpoint.from_directory("some_dir")
get_session().report(..., checkpoint)
# User's responsibility to clean up some_dir (and we need to tell them when...). But by this time, we may not have finished uploading to cloud yet...

Reply:
Yes, that is correct. Although (2) may not be that straightforward with current API (as opposed to the with tune.checkpoint_dir(iter)), where iter is used as the index into sub-directory.
Here is one proposal:
1. user can always write to "whatever_directory" under the current working directory (which is changed at the beginning)
2. upon report() returned, our framework makes sure that the move is completed to move "whatever_directory" to "checkpoint_{iter}" directory. The operation should be cheap.
3. Now, our framework can manage this "checkpoint_{iter}" directory the same way that we do to Tune checkpoints today.
4. Thus, the restriction on syncing semantics can be removed. 

cc @krfricke to cross check with syncer logic.

Reply:
Yeah :) 
Although i would like this contract to decouple from how our internal syncing logic is working. It can be sync or async, but that should be a decision weighing consistency and performance, and not because of the restrictions here.
So if you look up for the proposal I put together, I think it will work regardless sync or async :) 
Plus Jun always want our syncing to be non blocking. So let's keep it open.